From 1ac2a2fb621f06d27c77f172350f20f3719f0968 Mon Sep 17 00:00:00 2001
From: Volodymyr Bendiuga <volodymyr.bendiuga@westermo.se>
Date: Thu, 24 May 2018 15:23:13 +0200
Subject: [PATCH 31/38] [WMO]route/cls: return only one instance of each
 clasifier
Organization: Westermo Network Technologies AB

Kernel seems to return few instances of each classifier,
and this becomes problematic when we want to delete or add new
classifiers. This patch makes only one instance of each classifier
to be written to cache. This makes handling of cache very easy
and smooth.

Signed-off-by: Volodymyr Bendiuga <volodymyr.bendiuga@westermo.se>
Signed-off-by: Joachim Nilsson <troglobit@gmail.com>
---
 lib/route/cls.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/lib/route/cls.c b/lib/route/cls.c
index 1802ad6..25618a4 100644
--- a/lib/route/cls.c
+++ b/lib/route/cls.c
@@ -464,6 +464,20 @@ static int cls_msg_parser(struct nl_cache_ops *ops, struct sockaddr_nl *who,
 	if (cls->c_protocol)
 		cls->ce_mask |= CLS_ATTR_PROTOCOL;
 
+	/* TEMP: THIS IS NOT NICE, FIGURE OUT SMTHNG BETTER */
+	if (!cls->c_handle) {
+	        err = 0;
+		goto errout;
+	}
+
+	if (cls->c_handle >= 0x8000) {
+	        if ((cls->c_handle & 0x0800) == 0) {
+		        err = 0;
+			goto errout;
+		}
+	}
+	/* ********************************* */
+
 	err = pp->pp_cb(OBJ_CAST(cls), pp);
 errout:
 	rtnl_cls_put(cls);
-- 
2.25.1

